### ISSUE TITLE
[a11y]: NumberInput: use builtin increment/decrement pseudoelements?

### ISSUE THREAD

**Issue by wkeese** [carbon-design-system/carbon#18078]
### Package

@carbon/react

### Browser

Chrome

### Operating System

MacOS

### Package version

1.69.0

### React version

18

### Automated testing tool and ruleset

VoiceOver

### Description

[NumberInput](https://react.carbondesignsystem.com/?path=/story/components-numberinput--default) explicitly renders the increment and decrement buttons using the `<button>` element.

```html
<div class="cds--number__controls">
  <button aria-label="Decrement number" class="cds--number__control-btn down-icon" tabindex="-1" title="Decrement number" type="button">...</button>
  <div class="cds--number__rule-divider"></div>
  <button aria-label="Increment number" class="cds--number__control-btn up-icon" tabindex="-1" title="Increment number" type="button">../button>
  <div class="cds--number__rule-divider">
</div>
```
These buttons are useless to keyboard users since keyboard users can increment/decrement the value with the up and down arrow keys.

The buttons have `tabindex=-1` so they aren't in the tab-navigation, but they still have various accessibility issues.

# Acccessibility issues

(1) Ambiguity

If we have multiple `NumberInput` components they will essentially render like this:

```html
<label for=p>Precision</label>
<input id=p>
<button>increment</button>
<button>decrement</button>
<label for=s>Scale</label>
<input id=s>
<button>increment</button>
<button>decrement</button>
```

I think a human can guess which inputs each button belongs to, but it's not entirely clear.  It's also annoying to write unit tests if (for some reason) the unit tests wants to (for example) click the decrement button of the second NumberInput.

(2) VoiceOver

Although JAWS and NVDA will skip over the buttons, VoiceOver will not.   This seems suboptimal for VoiceOver users that have no interest in those buttons.

## Possible solutions

(3) How about setting `aria-hidden` around the buttons?
 
I don't think this works, it's problematic to have interactive controls inside an `aria-hidden` block.

(4) How about wrapping the whole thing in a `<fieldset>`?

```html
<fieldset>
    <legend>Precision</legend>
    <label for=p>Precision</label>
    <input id=p>
    <button>increment</button>
    <button>decrement</button>
</fieldset>
```

I don't like that either, it's problematic because the label is repeated twice.

(5) Styling buttons

Therefore, my suggestion is to the browser's native buttons, but style them how carbon wants them, i.e. NumberInput simply becomes

```html
<input type=number class=class="cds--number__input"/>
```

And then all the magic happens in `cds--number__input`.

See https://stackoverflow.com/questions/45396280/customizing-increment-arrows-on-input-of-type-number-using-css.

It sure looks possible.

### Reproduction/example

https://react.carbondesignsystem.com/?path=/story/components-numberinput--default

### Steps to reproduce

N/A

### Suggested Severity

None

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [x] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

**Comment by tay1orjones** [comment]
Another concern to add to the mix is the `type="text"` feature 
* https://github.com/carbon-design-system/carbon/pull/19471 

Also related: 
* https://github.com/carbon-design-system/carbon/issues/16529

We could reassess this whole situation with the Carbon Accessibility Guild and our manual testers to confirm expectations.

**Comment by sunny-kukkar-ibm** [comment]
@riddhybansal  will take it up with accessibility team to get this review.

**Comment by chrisgrimley** [comment]
Hi @wkeese, I've picked up this ticket since it still has the 'needs accessibility info' status attached. I had a quick look at the issues you raised. It's possible I'm missing some details or context, but from what I can see, the NumberInput has been implemented as expected. As you point out, the buttons are hidden with tabindex="-1" since keyboard users can use the keyboard controls to increment/decrement. This allows keyboard users full functionality while limiting tab stops.

The issue you raised around VoiceOver announcing the buttons does not happen for me. They remain hidden.

With regards the second issue, I agree that it could be confusing if there are multiple inputs beside each other, but my suggestion (and best practice) would be to add additional context in the label and aria-label to make it clear to AT users what each input applies to.

Please let me know if there are other considerations I've overlooked and have over-simplified the issue. Thanks.

fyi @tay1orjones

**Comment by wkeese** [comment]
> The issue you raised around VoiceOver announcing the buttons does not happen for me. They remain hidden.

You probably weren't testing on iOS.  It still happens with the latest code and you can hear it on https://react.carbondesignsystem.com/?path=/story/components-numberinput--default:

https://github.com/user-attachments/assets/6e927191-2e5f-4205-822c-319037f24a8f

> With regards the second issue, I agree that it could be confusing if there are multiple inputs beside each other, but my suggestion (and best practice) would be to add additional context in the label and aria-label to make it clear to AT users what each input applies to.

The second issue was VoiceOver.  It sounds like you are talking about the first issue.  But you misunderstood it.  The problem is not the labels on the inputs, but the labels on the buttons:

```html
<label for=p>Precision</label>
<input id=p>
<button>increment</button>
<button>decrement</button>
<label for=s>Scale</label>
<input id=s>
<button>increment</button>
<button>decrement</button>
```

The only way to set the button labels is by specifying a custom `translateWithId()` method for each NumberInput.   It's possible, but kludgey.

----

IMO though there's also another "issue", besides the accessibility, which is that we should never add a bunch of code for no reason.  (In this case, there's no reason to add those `<button>` nodes or the corresponding JS.)

**Comment by chrisgrimley** [comment]
Hi @wkeese, I was focusing on MacOS since you stated that the issue was on MacOS using Chrome. I'm pretty sure that announcing the buttons on iOS using VoiceOver with swipe gestures is expected. Wouldn't adding aria-label on the button (label was a typo above) make the announcement clear?

**Comment by wkeese** [comment]
>  I was focusing on MacOS since you stated that the issue was on MacOS using Chrome. 

Woops, not sure why I wrote MacOS instead of iOS, except that the first issue affects everyone.

> I'm pretty sure that announcing the buttons on iOS using VoiceOver with swipe gestures is expected.

I'm skeptical of that.  I assume most people would expect Carbon's NumberInput to work like https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/input/number.

> Wouldn't adding aria-label on the button (label was a typo above) make the announcement clear?

I think it would be clearer if each button had its own unique label, but it would still be a bad UX, and it would also be annoying for developers to have to specify (and translate) three labels for every `NumberInput`.

**Comment by chrisgrimley** [comment]
Ok, thanks for the additional info Bill. This might be a good topic to bring to the Accessibility Office Hours session today (3pm Dublin time). If you can't make it, I'd be happy to raise with the team and get their thoughts.

**Comment by tombrunet** [comment]
It's a little odd, but I'm tempted to say that those increment / decrement elements should be `span`. These get into that strange space there these are mouse affordances. Accessibility is generally focused on providing keyboard alternatives to visual/mouse functions, but in this case, those behaviors are provided by the native widget - so we're trying to provide the sighted/mouse user an equivalent to the native keyboard behaviors. Since we're hiding those visual aspects of the native widget, we have to reproduce them. I'd go with an aria-hidden span that's not in the tab order.

**Comment by wkeese** [comment]
(edited)

I'm not sure what switching from `<button>` to `<span>` would do.  I don't think it improves the iOS VoiceOver experience.

Maybe you misunderstood me, because I'm saying we keep the display and behavior EXACTLY THE SAME (except for screen reader users), and just change the IMPLEMENTATION DETAILS.  I.E. reduce the markup down to a single `<input>`, but display the buttons just like before by using CSS.

Having said that though, I just spent some time playing with the `::-webkit-inner-spin-button` pseudo-element, and it seems like there's no way to style it like how NumberInput currently appears.   I thought I could set the styling and content for each spinner button, something like this:

```css
input[type=number]::webkit-inner-spin-button {
   &:first-child {
        ...
        &::after { content: "-" }
   }
   &:last-child {
        ...
        &::after { content: "+" }
   }
}
```

(And then equivalent code for Firefox.)

However, now it seems like `webkit-inner-spin-button` is just the styling on a wrapper around both buttons, and there's no way to style the buttons themselves.

**Comment by tombrunet** [comment]
Right, when I read the issue you noted, it looked like the only standard option was to hide them completely and make your own, hence the suggestion of an aria-hidden span. The span shouldn't be recognized by Voiceover as a widget.

**Comment by wkeese** [comment]
> Right, when I read the issue you noted, it looked like the only standard option was to hide [the spinner buttons] completely

Disappointing.

> The span shouldn't be recognized by Voiceover as a widget.

Not sure what you mean by "recognized".  Doesn't VoiceOver stop on everything, even plain text?   That's how it's different than using JAWS etc.

PS: I'm talking about VoiceOver on iOS.

**Comment by wkeese** [comment]
Anyway, sounds like my idea is impossible currently.  Maybe it will be possible in the future.  But I'll close the ticket for now.
